src/main/java/com/google/devtools/build/lib/remote/merkletree/MerkleTreeUploader.java

```diff
@@ -0,0 +1,40 @@
+package com.google.devtools.build.lib.remote.merkletree;
+
+import build.bazel.remote.execution.v2.Digest;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.devtools.build.lib.actions.cache.VirtualActionInput;
+import com.google.devtools.build.lib.remote.common.RemoteActionExecutionContext;
+import com.google.devtools.build.lib.remote.common.RemotePathResolver;
+import com.google.devtools.build.lib.vfs.Path;
+import java.io.IOException;
+
+/** The basic cache operations needed to upload a {@link MerkleTree} and its associated blobs. */
+public interface MerkleTreeUploader {
+  /** Uploads an in-memory blob to the remote cache. */
+  ListenableFuture<Void> upload(RemoteActionExecutionContext context, Digest digest, byte[] data);
```

Comment:
nit: rename upload methods accordingly, e.g. `uploadBlob`, instead of using overloads.